File: ports/go/src/core/graphics/bcspainter.go
package graphics

import (
	"sync"
	core "github.com/robertpelloni/btk-go/ports/go/src/core/kernel"
)

type BcsPainter struct {
	*core.BcsObject
	canvas *BcsCanvas
	active bool
	mu     sync.Mutex
}

func NewBcsPainter(parent *core.BcsObject) *BcsPainter {
	return &BcsPainter{
		BcsObject: core.NewBcsObject(parent),
		active:    false,
	}
}

func (p *BcsPainter) Begin(canvas *BcsCanvas) bool {
	p.mu.Lock()
	defer p.mu.Unlock()
	if p.active {
		return false
	}
	p.canvas = canvas
	p.active = true
	return true
}

func (p *BcsPainter) End() {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.active = false
	p.canvas = nil
}

File: ports/rust/src/core/graphics/bcs_painter.rs
use std::sync::{Arc, Mutex};
use crate::core::kernel::bcs_object::BcsObject;
use super::bcs_canvas::BcsCanvas;

pub struct BcsPainter {
    pub base: Arc<BcsObject>,
    canvas: Mutex<Option<Arc<BcsCanvas>>>,
    active: Mutex<bool>,
}

impl BcsPainter {
    pub fn new(parent: Option<Arc<BcsObject>>) -> Arc<Self> {
        let base = if let Some(p) = parent {
            BcsObject::new_with_parent(p)
        } else {
            BcsObject::new()
        };

        Arc::new(Self {
            base,
            canvas: Mutex::new(None),
            active: Mutex::new(false),
        })
    }

    pub fn begin(&self, canvas: Arc<BcsCanvas>) -> bool {
        let mut active = self.active.lock().unwrap();
        if *active {
            return false;
        }
        *self.canvas.lock().unwrap() = Some(canvas);
        *active = true;
        true
    }

    pub fn end(&self) {
        let mut active = self.active.lock().unwrap();
        *active = false;
        *self.canvas.lock().unwrap() = None;
    }
}
